mathics/eval/arithmetic.py

```diff
@@ -354,7 +372,7 @@ def segregate_numbers(
 # TODO: Annotate me
 def segregate_numbers_from_sorted_list(
     *elements: BaseElement,
-) -> Tuple[List[Number], List[BaseElement]]:
+) -> Tuple[List[BaseElement], List[BaseElement]]:
```

Comment:
The point of this function is that the first element of the output tuple should be `Number`s. But of course, they are also `BaseElement`s.

Reply:
Again, there is a confusion about how type annotation works. It is not about what you can prove prove what the type system sees and its limitations. 

In the return on line 385 the type system is not smart enough reason that if there is an element that is not a number an early return is taken. 

From its less sophisticated level of understanding, the type of `*elements` has to match the return type of the first argument. The code can't guarantee that the type of *elements is a Number which is why it has to check in the loop.

Reply:
> In the return on line 385 the type system is not smart enough reason that if there is an element that is not a number an early return is taken.

If the program reaches the line 385, is because all the items of `elements` are `Number`, so indeed, the return type is correct, but OK. In that case, what about the line 361 of the `segregate_numbers` function?

Reply:
> > In the return on line 385 the type system is not smart enough reason that if there is an element that is not a number an early return is taken.
> 
> If the program reaches the line 385, is because all the items of `elements` are `Number`, so indeed, the return type is correct, but OK. In that case, what about the line 361 of the `segregate_numbers` function?

That one mypy and other type checkers think is okay because there _that list never had a type_. It was implicitly defined `[]` so it matches anything. 

Here is what mypy thinks of arithmetic.py master: 

[mypy.log](https://github.com/Mathics3/mathics-core/files/11183533/mypy.log)

Reply:
OK, the linter rules. Merge then when you feel this is ready.